FEAT: Exposed the array variable's DotDu value via coupledArrayDotDu

NateAM opened the pull request:
## Discussion

The current implementation of array variable's duDotDu returns a reference to a VariableValue because the derivatives are the same for all of the variables (due to identical time integration scheme). Consequently, the implementation of coupledArrayDotDu also returns a reference to a VariableValue.

closes #25833

NateAM committed:
FEAT: Exposed the array variable's DotDu value via coupledArrayDotDu

The current implementation of array variable's duDotDu returns a  rather than an  which seems to imply that the derivatives are the same for all of the variables. This may or may not be desired behavior.

closes #25833

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1840209/) on 526db53 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format bdad35fcdd7aa0bf79eaaa60e3372853c09c9023`

NateAM committed:
MAINT: Updated to match style guidelines

lindsayad commented:
Hi @NateAM, thanks for your pull request! let me know if you have any questions during the process. Looks like you'll need to apply some clang formatting. Hopefully the message is clear, but let me know if you have any trouble with that

lindsayad reviewed (commented):
It would be nice to add a test for this. If you cannot come up with one, then I probably can

NateAM commented:
> It would be nice to add a test for this. If you cannot come up with one, then I probably can

I could make something though I _think_ I would need to implement a new kernel to do it (an `ArrayCoupledTimeDereivative`). Maybe that goes beyond the scope of the PR?

YaqiWang commented:
I think `duDotDu` only depends on the time integration scheme, thus for all components of an array variable, they are the same. You are returning `VariableValue`, which seems contradict with what you said in the message?

NateAM commented:
> I think `duDotDu` only depends on the time integration scheme, thus for all components of an array variable, they are the same. You are returning `VariableValue`, which seems contradict with what you said in the message?

I've edited the original message to make my meaning (hopefully!) clear.

lindsayad commented:
I might say:

The current implementation of array variable's duDotDu returns a reference to a VariableValue because the derivatives are the same for all of the variables (due to identical time integration scheme). Consequently, the implementation of coupledArrayDotDu also returns a reference to a VariableValue.

moosebuild commented:
Job [Documentation](https://civet.inl.gov/job/1840292/) on 8667757 wanted to post the following:

View the site [here](https://mooseframework.inl.gov/docs/PRs/25837/site)

This comment will be updated on new commits.

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1840310/) on 8667757 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>bdad35</th>
      <th colspan=4>#25837 866775</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.63%</td>
      <td>85.62%</td>
      <td>-0.01%</td>
      <td>0.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>92354</td>
      <td>92350</td>
      <td>-4</td>
      <td>0</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15499</td>
      <td>15516</td>
      <td>+17</td>
      <td>13</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25837/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25837/coverage/framework)


## Modules coverage

Coverage did not change

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/geochemistry)
- [`heat_transfer`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/heat_transfer)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25837/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 0.00% is less than the suggested 90.0%



This comment will be updated on new commits.

NateAM committed:
FEAT: Added an array coupled time derivative kernel

NateAM committed:
TST: Added tests for the array coupled time derivative kernel's jacobian

NateAM committed:
MAINT: Corrected failure in the tests

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1887808/) on f542897 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5dfe9524b833c330429113a972e3984fff97c3c2`

NateAM committed:
MAINT: Updating style

NateAM commented:
@lindsayad I added a new kernel (`ArrayCoupledTimeDerivative`) to test the additions proposed here. Hopefully this helps get this PR merged.

NateAM committed:
MAINT: Updating the string so it compiles with METHOD=dbg

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1888167/) on 2907ca6 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 5dfe9524b833c330429113a972e3984fff97c3c2`

NateAM committed:
MAINT: Updating style

lindsayad commented:
thanks @NateAM !

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/tests:
```diff
@@ -0,0 +1,29 @@
+[Tests]
+  [./jac_1d]
```
```suggestion
  [jac_1d]
```

NateAM replied:
Done

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/tests:
```diff
@@ -0,0 +1,29 @@
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 1d grid'
+  [../]
+  [./jac_2d]
```
```suggestion
  []
  [jac_2d]
```

NateAM replied:
Done

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/tests:
```diff
@@ -0,0 +1,29 @@
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 2d grid'
+  [../]
+  [./jac_3d]
```
```suggestion
  []
  [jac_3d]
```

NateAM replied:
Done

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/tests:
```diff
@@ -0,0 +1,29 @@
+    ratio_tol = 1e-7
+    difference_tol = 1e-6
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 3d grid'
+  [../]
```
```suggestion
  []
```

NateAM replied:
Done

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/tests:
```diff
@@ -0,0 +1,29 @@
+[Tests]
+  [./jac_1d]
```
We will need a `design` and `issues` for this. This is why the documentation target is failing. If you need any help on how to add those, just let me know!

NateAM replied:
@lindsayad I think I've gotten the `issues` and `design` sorted out. I basically copied `CoupledTimeDerivative` word for word with some small changes. Let me know if that does what you need.

NateAM committed:
MAINT: Adding requests from reviewers

- typo fix
- whitespace fix
- more const
- remove unnecessary virtual
- remove use of `use_displaced_mesh = false`
- update requirement wording
- hit formatting

Co-authored-by: Alex Lindsay <alexander.lindsay@inl.gov>

lindsayad commented on framework/doc/content/source/kernels/ArrayCoupledTimeDerivative.md:
```diff
@@ -0,0 +1,48 @@
+The `ArrayCoupledTimeDerivative` kernel is very similar to the
+[`CoupledTimeDerivative`](/CoupledTimeDerivative.md) kernel with the
+exception that it works for array variables rather than scalars.
+The strong form on the the domain $\Omega$ is
```
```suggestion
The strong form on the domain $\Omega$ is
```

lindsayad commented on framework/doc/content/source/kernels/ArrayCoupledTimeDerivative.md:
```diff
@@ -0,0 +1,48 @@
+variable `v` that the time derivative operator acts upon. Example syntax can be
+found in the kernel block below:
+
+!listing
+test/tests/kernels/array_coupled_time_derivative/test_jacobian.i
+block=Kernels label=false
```
This doesn't show up correctly in the documentation (nor does it for the `CoupledTimeDerivative.md` file). I think the issue might be the line breaks. Can you try removing the line breaks?

NateAM replied:
I've tried to mimic one that seems to work correctly.

lindsayad commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -1084,6 +1084,17 @@ class Coupleable
+   * @return Reference to a ArrayVariableValue containing the time derivative of the coupled
+   * variable
+   */
+
```
```suggestion
```

lindsayad commented on framework/include/interfaces/Coupleable.h:
```diff
@@ -1084,6 +1084,17 @@ class Coupleable
+   * variable
+   */
+
+  virtual const VariableValue & coupledArrayDotDu(const std::string & var_name,
```
```suggestion
  const VariableValue & coupledArrayDotDu(const std::string & var_name,
```

lindsayad replied:
I don't know why the other ones are marked `virtual`.

lindsayad commented on framework/include/kernels/ArrayCoupledTimeDerivative.h:
```diff
@@ -0,0 +1,32 @@
+  virtual RealEigenVector computeQpJacobian() override;
+  virtual RealEigenMatrix computeQpOffDiagJacobian(const MooseVariableFEBase & jvar) override;
+
+  const ArrayVariableValue & _v_dot;
+  const VariableValue & _dv_dot;
+  const unsigned int _v_var;
```
Some short doxygen would be good

NateAM replied:
Done

lindsayad commented on framework/src/interfaces/Coupleable.C:
```diff
@@ -1427,6 +1427,31 @@ Coupleable::coupledDotDotDu(const std::string & var_name, unsigned int comp) con
+const VariableValue &
+Coupleable::coupledArrayDotDu(const std::string & var_name, unsigned int comp) const
+{
+  const auto * var = getArrayVar(var_name, comp);
```
```suggestion
  const auto * const var = getArrayVar(var_name, comp);
```

lindsayad commented on framework/src/kernels/ArrayCoupledTimeDerivative.C:
```diff
@@ -0,0 +1,60 @@
+  InputParameters params = ArrayKernel::validParams();
+  params.addClassDescription(
+      "Time derivative Array Kernel that acts on a coupled variable. Weak form: "
+      "$(\\psi_i, \\frac{\\partial v_h}{\\partial t})$. The coupled varialbe and"
```
```suggestion
      "$(\\psi_i, \\frac{\\partial v_h}{\\partial t})$. The coupled variable and"
```

lindsayad commented on framework/src/kernels/ArrayCoupledTimeDerivative.C:
```diff
@@ -0,0 +1,60 @@
+void
+ArrayCoupledTimeDerivative::computeQpResidual(RealEigenVector & residual)
+{
+
```
```suggestion
```

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/test_jacobian.i:
```diff
@@ -0,0 +1,93 @@
+    type = ArrayCoupledTimeDerivative
+    variable = u
+    v = v
+    use_displaced_mesh = false
+  []
+  [u_time_derivative]
+    type = ArrayTimeDerivative
+    variable = u
+    use_displaced_mesh = false
+  []
+  [u_diffusion]
+    type = ArrayDiffusion
+    variable = u
+    diffusion_coefficient = u_dc
+    use_displaced_mesh = false
+  []
+  [v_time_derivative]
+    type = ArrayTimeDerivative
+    variable = v
+    use_displaced_mesh = false
+  []
+  [v_diffusion]
+    type = ArrayDiffusion
+    variable = v
+    diffusion_coefficient = v_dc
+    use_displaced_mesh = false
```
```suggestion
  []
  [u_time_derivative]
    type = ArrayTimeDerivative
    variable = u
  []
  [u_diffusion]
    type = ArrayDiffusion
    variable = u
    diffusion_coefficient = u_dc
  []
  [v_time_derivative]
    type = ArrayTimeDerivative
    variable = v
  []
  [v_diffusion]
    type = ArrayDiffusion
    variable = v
    diffusion_coefficient = v_dc
```

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/test_jacobian.i:
```diff
@@ -0,0 +1,93 @@
+[]
+
+[Preconditioning]
+  [./smp]
```
```suggestion
  [smp]
```

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/test_jacobian.i:
```diff
@@ -0,0 +1,93 @@
+  [./smp]
+    type = SMP
+    full = true
+  [../]
```
```suggestion
  []
```

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/tests:
```diff
@@ -0,0 +1,35 @@
+    difference_tol = 1e-6
+    issues = '#25833'
+    design = 'kernels/ArrayCoupledTimeDerivative.md'
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 1d grid'
```
```suggestion
    requirement = 'The system shall be able to reproduce the exact Jacobian for a 1d grid for a coupled array time derivative kernel.'
```

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/tests:
```diff
@@ -0,0 +1,35 @@
+    difference_tol = 1e-6
+    issues = '#25833'
+    design = 'kernels/ArrayCoupledTimeDerivative.md'
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 2d grid'
```
```suggestion
    requirement = 'The system shall be able to reproduce the exact Jacobian for a 2d grid for a coupled array time derivative kernel.'
```

lindsayad commented on test/tests/kernels/array_coupled_time_derivative/tests:
```diff
@@ -0,0 +1,35 @@
+    difference_tol = 1e-6
+    issues = '#25833'
+    design = 'kernels/ArrayCoupledTimeDerivative.md'
+    requirement = 'We shall be able to reproduce the exact Jacobian for a 3d grid'
```
```suggestion
    requirement = 'The system shall be able to reproduce the exact Jacobian for a 3d grid for a coupled array time derivative kernel.'
```

lindsayad reviewed (changes requested):
great contribution! Just some small suggested changes

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1894698/) on efbde32 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2a397e13fdb83e5bf9f5a463f4196eacca670e42`

NateAM committed:
DOC: Modifying listing to try and fix display in documentation

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1894742/) on da6a92c wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2a397e13fdb83e5bf9f5a463f4196eacca670e42`

NateAM committed:
DOC: Added documentation for the new internal variables

moosebuild commented:
Job [Precheck](https://civet.inl.gov/job/1894830/) on 915e445 wanted to post the following:

Your code requires style changes.  

A patch was auto generated and copied [here](https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch)  
You can directly apply the patch by running, in the top level of your repository:

`curl -s https://mooseframework.inl.gov/docs/PRs/25837/clang_format/style.patch | git apply -v`  

Alternatively, with your repository up to date and in the top level of your repository:

`git clang-format 2a397e13fdb83e5bf9f5a463f4196eacca670e42`

NateAM committed:
MAINT: Updating format

lindsayad commented:
@NateAM FYI I just did a force push squashing some of those commits

lindsayad commented:
you are correct

lindsayad commented:
as stated earlier, failure is unrelated to this PR
